### PR TITLE
Feature/load first background

### PR DESCRIPTION
**Pull request by learnforward2023**
## Why

Fixes: Feature/load first background

https://studytogether.vn/code/basic/cpp/sky-birds-game/load-first-background

## What

- [Config CMakeLists.txt to load SDL2_image](https://github.com/learnforward2023/sky-birds-game/commit/946b29a006e4c976434d0038cbe29f686ded4052)
- [Stop using SDL_SetRenderDrawColor and use SDL_RenderCopy](https://github.com/learnforward2023/sky-birds-game/commit/fed8f8a39772d65932e79af146635c91ac387006)
- [Refactor: split rendering the background into a separate method](https://github.com/learnforward2023/sky-birds-game/commit/1345a68c9fb66b8152f1f3442b5ed3e6adca057c)
- [Refactor: stop hard code path, but using SDL_GetBasePath](https://github.com/learnforward2023/sky-birds-game/commit/618c760b1e11b338fababc23695e38c38b1f6a42)
- [Refactor: make images directory and push all images into it](https://github.com/learnforward2023/sky-birds-game/commit/a91455983cd853fa6bd4246b9f793b1f3394f341)

## Other Information

Website: https://studytogether.vn/ 🌟
Tiktok: [Lets Study Together](https://www.tiktok.com/@study2together) 🙏

### CODE REVIEW

**Comment by learnforward2022** [comment]
Here is a summary and analysis of the changes made to the code, along with suggestions, tips, and better practices to follow:

### CMakeLists.txt
- The `find_package(SDL2 REQUIRED)` is unchanged, reflecting that the project already made use of SDL2.
- Two new lines have been added.
  - `find_package(SDL2_image REQUIRED)` is to include SDL2_image, a library that allows for easy image loading in various formats.
  - `set(SDL2IMAGE_INCLUDE_DIRS /opt/homebrew/include/SDL2)` explicitly sets the include directory for SDL2_image. This is necessary for the compiler to find image-related headers.

#### Suggestions for CMakeLists.txt:
- Using `REQUIRED` in `find_package` will cause the configuration process to fail if SDL2_image is not found, which is good for highlighting missing dependencies.
- Consider using `SDL2_IMAGE_INCLUDE_DIRS` (generated automatically if `FindSDL2_image.cmake` is correct) rather than hardcoding the path to `/opt/homebrew/include/SDL2`, which may not be the include path on all systems.

### Images
- A new image file called `background.png` has been added to the images directory. This is typically done when you want to add visual assets to your game.

### main.cpp
- Two headers have been included:
  - `#include <SDL_image.h>` to use SDL2_image functions for loading images.
  - `#include <string>` because std::string is used to manage file paths.
- A new function `renderBackground` has been introduced:
  - It constructs the path for `background.png` by getting the base application path and appending the relative path to the image.
  - It loads the image using `IMG_Load`, converts it to a texture, and renders it on the screen.
  - The loaded surface is freed after creating the texture.
  - Error handling is included to catch if the image or texture cannot be loaded.
- In the `main` function:
  - The setting of the render draw color to black has been removed, probably assuming that it will be entirely covered by the background image.
  - `renderBackground` is called to draw the background image to the screen.

#### Suggestions for main.cpp:
- Wrap resource-loading code (like textures) in a class or function that automatically handles freeing resources. This is more robust and allows easy reuse.
- Be careful with paths. The use of "../images/background.png" assumes a specific directory structure that might not hold in all deployment scenarios.
- Consider using smart pointers, such as `std::unique_ptr` with a custom deleter, to automatically manage SDL resources.
- Check for memory leaks or resources not being freed when the game exits or the texture needs to be reloaded.
- Encapsulate game-related variables and functions into classes to maintain organized code and separate concerns.

### Tips and Tricks
- Always check the return value of SDL functions and handle the errors gracefully.
- When working with paths, use filesystem utilities (e.g., `std::filesystem` in C++17) to build paths in a platform-independent way.
- Use comments to explain non-obvious code, especially when it involves complex logic or workarounds.

### Alternative Solutions
- Load textures during initialization and store them for repeated use instead of loading and destroying them every frame, which is inefficient.

### General Advice
- Always profile your code to find bottlenecks and optimizations.
- Consider building a resource manager to handle the loading, caching, and unloading of game assets.
- Make use of consistent naming conventions and code organization practices to make the code more maintainable.

Remember that game development often includes refining and performance tuning, so always revisit and profile your code to ensure optimal performance.

#### This comment is created by OpenAI API. 👍
Website: [https://studytogether.vn](https://studytogether.vn) 🌟
Tiktok: [Lets Study Together](https://www.tiktok.com/@study2together) 🙏

File: main.cpp [hunk 1]
```diff
@@ -7,6 +9,35 @@ const int GAME_SCREEN_WIDTH = 1400;
...
+
+  // Create texture from surface pixels
+  SDL_Texture* backgroundTexture = SDL_CreateTextureFromSurface(screen, backgroundSurface);
+  if (backgroundTexture == nullptr) {
+    printf("Unable to create texture from %s! SDL Error: %s\n", backgroundPath.c_str(), SDL_GetError());
+
+    return;
+  }
+
+  SDL_FreeSurface(backgroundSurface); // Free the surface after creating the texture
```
Comment (learnforward2023) [inline main.cpp:34]: Tại sao chúng ta phải free `surface` sau khi tạo `texture`? 🤔

File: main.cpp [hunk 2]
```diff
@@ -7,6 +9,35 @@ const int GAME_SCREEN_WIDTH = 1400;
...
+  SDL_Rect backgroundRect = {0, 0, GAME_SCREEN_WIDTH, GAME_SCREEN_HEIGHT};
+
+  // Render the background
+  SDL_RenderCopy(screen, backgroundTexture, nullptr, &backgroundRect);
```
Comment (learnforward2023) [inline main.cpp:38]: - Tại sao bạn lại truyền `nullptr` vào đối số thứ 3 khi gọi phương thức `SDL_RenderCopy`? 🤔 
- Tại sao lại là `&backgroundRect`, mà không phải là `backgroundRect` 💭

File: main.cpp [hunk 3]
```diff
@@ -7,6 +9,35 @@ const int GAME_SCREEN_WIDTH = 1400;
...
+  // Create texture from surface pixels
+  SDL_Texture* backgroundTexture = SDL_CreateTextureFromSurface(screen, backgroundSurface);
+  if (backgroundTexture == nullptr) {
+    printf("Unable to create texture from %s! SDL Error: %s\n", backgroundPath.c_str(), SDL_GetError());
+
+    return;
+  }
+
+  SDL_FreeSurface(backgroundSurface); // Free the surface after creating the texture
+  SDL_Rect backgroundRect = {0, 0, GAME_SCREEN_WIDTH, GAME_SCREEN_HEIGHT};
```
Comment (learnforward2023) [inline main.cpp:35]: - Giải thích đoạn code này 🙏

File: main.cpp [hunk 4]
```diff
@@ -7,6 +9,35 @@ const int GAME_SCREEN_WIDTH = 1400;
...
 const char* GAME_TITLE = "Study Together - Sky Birds Game";
 
+void renderBackground() {
+  // Get the base path for the game application
+  std::string basePath = SDL_GetBasePath();
+  // Append the image to the base path
+  std::string backgroundPath = basePath.append("../images/background.png");
+
+  // Load Background image
+  SDL_Surface* backgroundSurface = IMG_Load(backgroundPath.c_str());
```
Comment (learnforward2023) [inline main.cpp:19]: - Tại sao lại phải `backgroundPath.c_str()`, không phải chỉ cần `backgroundPath ` là được sao?

File: main.cpp [hunk 5]
```diff
@@ -7,6 +9,35 @@ const int GAME_SCREEN_WIDTH = 1400;
 const int GAME_SCREEN_HEIGHT = 740;
 const char* GAME_TITLE = "Study Together - Sky Birds Game";
 
+void renderBackground() {
+  // Get the base path for the game application
+  std::string basePath = SDL_GetBasePath();
+  // Append the image to the base path
+  std::string backgroundPath = basePath.append("../images/background.png");
```
Comment (learnforward2023) [inline main.cpp:16]: - Tại sao lại phải `../images` mà không phải `images` luôn, còn phải `../` làm gì? 🤔

File: main.cpp [hunk 6]
```diff
@@ -8,6 +8,30 @@ const int GAME_SCREEN_WIDTH = 1400;
...
+    printf("Unable to create texture from %s! SDL Error: %s\n", "../images/background.bmp", SDL_GetError());
+
+    return;
+  }
+
+  SDL_FreeSurface(backgroundSurface); // Free the surface after creating the texture
+  SDL_Rect backgroundRect = {0, 0, GAME_SCREEN_WIDTH, GAME_SCREEN_HEIGHT};
+
+  // Render the background
+  SDL_RenderCopy(screen, backgroundTexture, nullptr, &backgroundRect);
```
Comment (learnforward2023) [inline main.cpp:38]: Đối tượng `screen` lấy ở đâu ra đây?

**Pull request merged**